File: Python/S0007-reverse-integer.py
# Given a 32-bit signed integer, reverse digits of an integer.
#
# Example 1:
# Input: 123
# Output: 321
#
#
# Example 2:
# Input: -123
# Output: -321
#
#
# Example 3:
# Input: 120
# Output: 21
#
#  Note:
# Assume we are dealing with an environment which could only store integers with
# in the 32-bit signed integer range: [−231, 231 − 1]. For the purpose of this problem,
# assume that your function returns 0 when the reversed integer overflows.

class Solution:
    def reverse(self, x: int) -> int:
        neg = 1
        if x < 0:
            neg = -1
            x *= -1
        out = 0
        while x > 0:
            rem = x % 10
            x //= 10
            out = out * 10 + rem

        if out >= 2**31-1 or out <= -2**31:
            return 0
        return out * neg

if __name__ == '__main__':
    assert Solution().reverse(123) == 321
    assert Solution().reverse(-123) == -321
    assert Solution().reverse(120) == 21
    assert Solution().reverse(123654) == 456321
    assert Solution().reverse(-123654) == -456321
    assert Solution().reverse(1534236469) == 0

File: Rust/S0007-reverse-integer.rs
/*
Given a 32-bit signed integer, reverse digits of an integer.

- Example 1:

Input: 123
Output: 321

- Example 2:

Input: -123
Output: -321

- Example 3:

Input: 120
Output: 21

Note:
Assume we are dealing with an environment which could only
store integers within the 32-bit signed integer range: [−2^31,  2^31 − 1].
For the purpose of this problem, assume that your function returns 0
when the reversed integer overflows.
*/
pub struct Solution {}

impl Solution {
	pub fn reverse(x: i32) -> i32 {
		let neg = if x < 0 {
			-1
		} else {
			1
		};

		let mut input = if x < 0 {
			-x
		} else {
			x
		} as i64;
		let mut out = 0i64;
		while input > 0 {
		    out *= 10;
		    out += &input % 10;
		    input /= 10;
		}
		if out > std::i32::MAX as i64 || out < std::i32::MIN as i64 {
			return 0;
		}
		return (out*neg) as i32;
	}
}

fn main() {
  assert_eq!(321, Solution::reverse(123));
  assert_eq!(-321, Solution::reverse(-123));
  assert_eq!(21, Solution::reverse(120));
  assert_eq!(0, Solution::reverse(1534236469));
  println!("Pass test cases!");
}

